pkg/controllers/v1alpha1/dataset/dataset_controller.go

```diff
@@ -201,13 +203,35 @@ func (r *DatasetReconciler) reconcileDatasetDeletion(ctx reconcileRequestContext
 		return utils.RequeueAfterInterval(time.Duration(10 * time.Second))
 	}
 
-	// 2. if there are datasets mounted this dataset, then requeue
-	if ctx.Dataset.Status.DatasetRef != nil && len(ctx.Dataset.Status.DatasetRef) > 0 {
-		ctx.Log.Error(err, "Failed to delete dataset because there are datasets mounted to it", "DatasetDeleteError", ctx,
-			"MountedDataset", ctx.Dataset.Status.DatasetRef)
-		r.Recorder.Eventf(&ctx.Dataset, v1.EventTypeWarning, common.ErrorDeleteDataset,
-			"Failed to delete dataset because there are datasets %s mounted to it", ctx.Dataset.Status.DatasetRef)
-		return utils.RequeueAfterInterval(10 * time.Second)
+	// 2. if there are datasets mounted this dataset, check reference dataset existence and requeue if there still has reference dataset
+	if len(ctx.Dataset.Status.DatasetRef) > 0 {
+		// 2.1 check reference dataset existence and remove not found dataset
+		datasetRefUpToDate, err := utils.RemoveNotFoundDatasetRef(r.Client, ctx.Dataset, ctx.Log)
+		if err != nil {
+			ctx.Log.Error(err, "Failed to update datasetRef", "DatasetDeleteError", ctx)
+			return utils.RequeueAfterInterval(time.Duration(10 * time.Second))
+		}
+
+		// 2.2 update dataset if datasetRef change
+		if !reflect.DeepEqual(datasetRefUpToDate, ctx.Dataset.Status.DatasetRef) {
+			datasetUpToDate := ctx.Dataset.DeepCopy()
+			datasetUpToDate.Status.DatasetRef = datasetRefUpToDate
+			if err := r.Status().Update(ctx, datasetUpToDate); err != nil {
+				ctx.Log.Error(err, "DatasetRef has changed but update failed", "DatasetDeleteError", datasetUpToDate)
+				return utils.RequeueAfterInterval(time.Duration(10 * time.Second))
+			}
+			ctx.Log.Info("Update dataset datasetRef successfully", "Before", ctx.Dataset.Status.DatasetRef, "After", datasetRefUpToDate)
+			ctx.Dataset = *datasetUpToDate
```

Comment:
Assigning *datasetUpToDate to ctx.Dataset is an anti patten to update ctx.Dataset through value passing. It may have side effects because any subsequent changes made to ctx.Dataset or *datasetUpToDate can potentially affect the other. This approach is not commonly used because it can introduce potential issues in code management and collaboration. A better practice is to keep ctx.Dataset unchanged and use datasetUpToDate instead.

Reply:
`datasetUpToDate` is a local variable and can not be used to remove finalizer. I think we can get the dataset after update dataset datasetRef. And if the dataset is up to date, we can continue following action. If it is not up to date, the dataset will be deleted in next reconcile.